sql/expression/function/json/json_search.go

```diff
@@ -0,0 +1,327 @@
+// must evaluate to a constant at compile time, not just at execution time. For example, if JSON_SEARCH() is used in a
+// prepared statement and the escape_char argument is supplied using a ? parameter, the parameter value might be
+// constant at execution time, but is not at compile time.
+//
+// https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-search
+type JSONSearch struct {
+	JSON     sql.Expression
+	OneOrAll sql.Expression
+	Search   sql.Expression
+	Escape   sql.Expression
+	Path     sql.Expression
+}
+
+var errOneOrAll = fmt.Errorf("the oneOrAll argument to json_search may take these values: 'one' or 'all'")
+var errBadEscape = fmt.Errorf("incorrect arguments to ESCAPE")
+
+var _ sql.FunctionExpression = &JSONSearch{}
+
+// NewJSONSearch creates a new NewJSONSearch function.
+func NewJSONSearch(args ...sql.Expression) (sql.Expression, error) {
+	switch len(args) {
+	case 3:
+		return &JSONSearch{
+			JSON:     args[0],
+			OneOrAll: args[1],
+			Search:   args[2],
+		}, nil
+	case 4:
+		return &JSONSearch{
+			JSON:     args[0],
+			OneOrAll: args[1],
+			Search:   args[2],
+			Escape:   args[3],
+		}, nil
+	case 5:
+		return &JSONSearch{
+			JSON:     args[0],
+			OneOrAll: args[1],
+			Search:   args[2],
+			Escape:   args[3],
+			Path:     args[4],
+		}, nil
+	default:
+		return nil, sql.ErrInvalidArgumentNumber.New("JSON_SEARCH", "3 to 5", len(args))
+	}
+}
+
+// FunctionName implements sql.FunctionExpression
+func (j *JSONSearch) FunctionName() string {
+	return "json_search"
+}
+
+// Description implements sql.FunctionExpression
+func (j *JSONSearch) Description() string {
+	return "path to value within JSON document."
+}
+
+// Resolved implements sql.Expression
+func (j *JSONSearch) Resolved() bool {
+	return j.JSON.Resolved() &&
+		j.OneOrAll.Resolved() &&
+		j.Search.Resolved() &&
+		(j.Escape == nil || j.Escape.Resolved()) &&
+		(j.Path == nil || j.Path.Resolved())
+}
+
+// String implements sql.Expression
+func (j *JSONSearch) String() string {
+	// TODO: maybe just don't print if escape/path are nil?
+	var escapeStr, pathStr string
+	if j.Escape == nil {
+		escapeStr = "NULL"
+	} else {
+		escapeStr = j.Escape.String()
+	}
+	if j.Path == nil {
+		pathStr = "NULL"
+	} else {
+		pathStr = j.Path.String()
+	}
+	return fmt.Sprintf("%s(%s, %s, %s, %s, %s)",
+		j.FunctionName(),
+		j.JSON.String(),
+		j.OneOrAll.String(),
+		j.Search.String(),
+		escapeStr,
+		pathStr,
+	)
+}
+
+// Type implements sql.Expression
+func (j *JSONSearch) Type() sql.Type {
+	return types.JSON
+}
+
+// IsNullable implements sql.Expression
+func (j *JSONSearch) IsNullable() bool {
+	return j.JSON.IsNullable() ||
+		j.OneOrAll.IsNullable() ||
+		j.Search.IsNullable() ||
+		(j.Escape != nil && j.Escape.IsNullable()) ||
+		(j.Path != nil && j.Path.IsNullable())
+}
+
+func jsonSearch(json interface{}, matcher *expression.LikeMatcher, currPath string, once bool) ([]string, bool) {
```

Comment:
(minor) would be nice to explain what the bool return param does, or... even just making it a named return param might be enough to make it more obvious that the boolean return param indicates whether the search stopped after the first match or not.

Reply:
added docstring to `jsonSearch()`